### DISCUSSION TITLE
Clarify recommended ESxxxx version

### DISCUSSION THREAD

**Discussion by davecardwell** [vercel/next.js#15873] [Ideas]
I noticed in https://github.com/vercel/next.js/pull/15772 that the title suggests ES2019 features will be polyfilled, but the description suggests ES2018.

The default for `lib` in [writeConfigurationDefaults.ts](https://github.com/vercel/next.js/blob/f5b186cb69039c354c6f02c76d5712dbf1170f71/packages/next/lib/typescript/writeConfigurationDefaults.ts#L25) seems to be ESNext if not already configured.

Looking through the `tsconfig.json` of the various with-x-typescript examples shows a mix in the `lib` setting, for example:

[with-typescript](https://github.com/vercel/next.js/blob/canary/examples/with-typescript/tsconfig.json): ES2017
[with-stripe-typescript](https://github.com/vercel/next.js/blob/canary/examples/with-stripe-typescript/tsconfig.json): ESNext

Is there a definitive ES version that the Next.js team are targeting or can at least recommend? I appreciate that individual projects can choose to add extra polyfills as needed, but I think it would help to be consistent and possibly document what the framework is targeting.

**Comment by AdamFielding** [comment]
I've been trying to determine what things are polyfilled by Next.js, and it's not at all obvious. As best I can tell, [this file](https://github.com/vercel/next.js/blob/canary/packages/next-polyfill-nomodule/src/index.js) includes all the polyfills that are provided on the condition that the browser returns true for the `nomodule` check.

So, the difficulty here is that the polyfills don't neatly fit into one of the [library groups](https://github.com/Microsoft/TypeScript/tree/master/lib) (ES2018 etc) that `tsconfig.json` allows you to specify. So Next.js can't easily say, "we're providing `Array.flatMap`, but not `Object.fromEntries`". 
Edit: This is slightly wrong, we actually can say things like:
```json
"lib": [
      "dom",
      "dom.iterable",
      "ES2018",
      "ES2019.Array"
],
```

Additionally, this also assumes that Next.js even knows what browsers we're targeting. Even if Next.js could specify exactly what polyfills it will provide to those `nomodule` browsers, how will it know what browsers the user is targeting? I don't want TypeScript complaining that `Object.fromEntries` doesn't exist if my application is only targeting up to date browsers.

This is all to say (from my cobbled together understanding) this is probably not a problem with a clear solution. What could help matters is to have some clearer documentation about how this works, and the reasoning behind it. Perhaps even just a link from [the docs](https://nextjs.org/docs/basic-features/supported-browsers-features) to [the list of polyfills](https://github.com/vercel/next.js/blob/canary/packages/next-polyfill-nomodule/src/index.js) we get from next.js
